File: tui/internal/mergepr/cigate.go
// Package mergepr is the merge-time gate for jonhill90/agent-tui,
// modelled directly on agent-supervisor's own working pattern
// (scripts/supervisor/merge-pr.sh + scripts/supervisor/ci_gate.py) per
// agent-tui#109's brief: two confirmed instances of a comment-verdict gate
// merged by its own author, unreviewed, within minutes (skills#255,
// agent-tui#107) mean the gate has to be enforced by the merge path
// itself, not documented as a convention a caller might skip.
//
// This package chains two gates, both fail-closed, before allowing a
// merge:
//
//  1. EvaluateCI -- CI must be green at the PR's CURRENT head SHA, ported
//     from ci_gate.py's own evaluate(): re-fetch headRefOid itself rather
//     than trust a caller-supplied SHA (a stale snapshot would let a since
//     -invalidated green reading through), read SHA-scoped check-runs and
//     commit-status endpoints rather than gh pr view's statusCheckRollup
//     (which carries no SHA a caller can compare), collapse re-runs to
//     latest-per-name the same way gh pr checks and GitHub's own PR UI do,
//     and refuse outright -- never "pass" -- when zero checks are reported
//     at all (absent is not pending).
//  2. internal/prverdict.Resolve -- an independent, current-head,
//     cross-lane APPROVE must be on record as a PR comment (this repo has
//     no ledger to resolve lane identity against tmux, so it trusts the
//     Author-Lane:/Review-Lane: trailers the same way internal/prverdict's
//     own doc comment explains).
//
// Merge (mergepr.go) is the only place that chains the two and calls
// `gh pr merge` -- exactly the role merge-pr.sh plays for agent-supervisor,
// stated the same way in this repo's own AGENTS.md: this is now THE way to
// merge a PR here, not a fourth optional tool alongside `gh pr merge` run
// by hand.
package mergepr

import (
	"encoding/json"
	"fmt"
	"sort"

	"github.com/jonhill90/agent-tui/internal/prverdict"
)

// Runner is the same seam shape as prverdict.Runner and internal/board's
// GitHubRunner -- kept as an alias, not a fresh named type, because this
// package always calls `gh` through the exact same execution path
// prverdict.ExecRunner already provides; a second Runner type here would
// buy nothing but an extra conversion at every call site.
type Runner = prverdict.Runner

// greenConclusions mirrors ci_gate.py's GREEN_CONCLUSIONS exactly:
// "neutral" and "skipped" both count as passing the same way GitHub's own
// PR UI treats them, not just "success".
var greenConclusions = map[string]bool{
	"success": true,
	"neutral": true,
	"skipped": true,
}

// CIDecision is EvaluateCI's whole answer -- mirrors prverdict.Decision's
// own pattern of a typed, exhaustive result rather than a bare bool, so a
// caller can log WHY as easily as WHETHER.
type CIDecision string

const (
	// CIAllow: every check reported for the current head SHA is green.
	CIAllow CIDecision = "allow"
	// CIRefuse: red, pending, missing entirely, or unreadable -- all the
	// same decision, same as ci_gate.py's own "absent is not pending".
	CIRefuse CIDecision = "refuse"
)

// CIResult is EvaluateCI's return value -- SHA is always populated once
// the head commit was successfully resolved, even on a refuse, so a
// caller (Merge) can still report which commit it evaluated.
type CIResult struct {
	Decision CIDecision
	SHA      string
	Reason   string
}

type prHeadView struct {
	HeadRefOid string `json:"headRefOid"`
}

type checkRun struct {
	Name        string `json:"name"`
	Status      string `json:"status"`
	Conclusion  string `json:"conclusion"`
	HeadSHA     string `json:"head_sha"`
	CompletedAt string `json:"completed_at"`
	StartedAt   string `json:"started_at"`
}

type checkRunsResponse struct {
	CheckRuns []checkRun `json:"check_runs"`
}

type commitStatus struct {
	Context string `json:"context"`
	State   string `json:"state"`
}

type commitStatusResponse struct {
	Statuses []commitStatus `json:"statuses"`
}

// EvaluateCI is the CI gate: green or refuse, for repo#number's CURRENT
// head SHA, re-fetched here rather than accepted as a parameter -- see
// this file's own doc comment for why a caller-supplied SHA would
// reintroduce the bug ci_gate.py's own doc comment names (check green for
// an older commit than head, because a push raced past a stale snapshot).
func EvaluateCI(run Runner, repo string, number int) CIResult {
	sha, err := headSHA(run, repo, number)
	if err != nil {
		return CIResult{Decision: CIRefuse, Reason: fmt.Sprintf("could not read PR head: %s", err)}
	}

	runs, err := fetchCheckRuns(run, repo, sha)
	if err != nil {
		return CIResult{Decision: CIRefuse, SHA: sha, Reason: fmt.Sprintf("could not read checks for %s: %s", sha, err)}
	}
	statuses, err := fetchCommitStatus(run, repo, sha)
	if err != nil {
		return CIResult{Decision: CIRefuse, SHA: sha, Reason: fmt.Sprintf("could not read checks for %s: %s", sha, err)}
	}

	for _, r := range runs {
		if r.HeadSHA != sha {
			return CIResult{
				Decision: CIRefuse,
				SHA:      sha,
				Reason:   fmt.Sprintf("check run %s reports head_sha %q, not PR head %s", r.Name, r.HeadSHA, sha),
			}
		}
	}

	if len(runs) == 0 && len(statuses) == 0 {
		return CIResult{Decision: CIRefuse, SHA: sha, Reason: fmt.Sprintf("no checks reported for %s", sha)}
	}

	var failing []string
	for _, r := range latestPerName(runs) {
		if !(r.Status == "completed" && greenConclusions[r.Conclusion]) {
			failing = append(failing, describeFailingRun(r))
		}
	}
	for _, s := range statuses {
		if s.State != "success" {
			name := s.Context
			if name == "" {
				name = "?"
			}
			failing = append(failing, name)
		}
	}

	if len(failing) > 0 {
		return CIResult{
			Decision: CIRefuse,
			SHA:      sha,
			Reason:   fmt.Sprintf("check(s) not green at %s: %s", sha, joinComma(failing)),
		}
	}

	return CIResult{Decision: CIAllow, SHA: sha, Reason: fmt.Sprintf("all checks green at %s", sha)}
}

func headSHA(run Runner, repo string, number int) (string, error) {
	out, err := run([]string{
		"pr", "view", fmt.Sprintf("%d", number),
		"--repo", repo,
		"--json", "headRefOid",
	})
	if err != nil {
		return "", err
	}
	var view prHeadView
	if err := json.Unmarshal(out, &view); err != nil {
		return "", fmt.Errorf("decode gh pr view: %w", err)
	}
	if view.HeadRefOid == "" {
		return "", fmt.Errorf("gh did not return a head SHA for this PR")
	}
	return view.HeadRefOid, nil
}

func fetchCheckRuns(run Runner, repo, sha string) ([]checkRun, error) {
	out, err := run([]string{
		"api", fmt.Sprintf("repos/%s/commits/%s/check-runs", repo, sha),
	})
	if err != nil {
		return nil, err
	}
	var resp checkRunsResponse
	if err := json.Unmarshal(out, &resp); err != nil {
		return nil, fmt.Errorf("decode check-runs: %w", err)
	}
	return resp.CheckRuns, nil
}

func fetchCommitStatus(run Runner, repo, sha string) ([]commitStatus, error) {
	out, err := run([]string{
		"api", fmt.Sprintf("repos/%s/commits/%s/status", repo, sha),
	})
	if err != nil {
		return nil, err
	}
	var resp commitStatusResponse
	if err := json.Unmarshal(out, &resp); err != nil {
		return nil, fmt.Errorf("decode commit status: %w", err)
	}
	return resp.Statuses, nil
}

// latestPerName collapses re-runs the same way gh pr checks and GitHub's
// own PR UI do: a re-run of `test` after a flaky failure leaves both the
// old failing run and the new passing one in the API response, and only
// the newest run of each check name counts. Sort key mirrors ci_gate.py's
// _run_sort_key: completed_at, falling back to started_at for a run still
// in flight -- both are ISO-8601 UTC ("...Z"), so lexical order matches
// chronological order without parsing timestamps.
func latestPerName(runs []checkRun) []checkRun {
	latest := map[string]checkRun{}
	for _, r := range runs {
		current, ok := latest[r.Name]
		if !ok || sortKey(r) >= sortKey(current) {
			latest[r.Name] = r
		}
	}
	names := make([]string, 0, len(latest))
	for name := range latest {
		names = append(names, name)
	}
	sort.Strings(names)
	out := make([]checkRun, 0, len(names))
	for _, name := range names {
		out = append(out, latest[name])
	}
	return out
}

func sortKey(r checkRun) string {
	if r.CompletedAt != "" {
		return r.CompletedAt
	}
	return r.StartedAt
}

// describeFailingRun mirrors ci_gate.py's _describe_failing_run: a merge
// gate that only ever printed "not green" makes `cancelled` (will never go
// green without a re-run) read identically to `in_progress` (may still go
// green on its own) -- whoever reads this refusal needs that distinction
// spelled out even though the decision itself is the same either way.
func describeFailingRun(r checkRun) string {
	if r.Status == "completed" && r.Conclusion == "cancelled" {
		return fmt.Sprintf("%s (cancelled -- will not go green without a re-run)", r.Name)
	}
	if r.Status != "completed" {
		status := r.Status
		if status == "" {
			status = "unknown"
		}
		return fmt.Sprintf("%s (%s -- may still go green)", r.Name, status)
	}
	conclusion := r.Conclusion
	if conclusion == "" {
		conclusion = "unknown"
	}
	return fmt.Sprintf("%s (%s)", r.Name, conclusion)
}

func joinComma(items []string) string {
	out := ""
	for i, item := range items {
		if i > 0 {
			out += ", "
		}
		out += item
	}
	return out
}

File: reference/scripts/supervisor/ci_gate.py
"""CI as a gate, not a badge (agent-supervisor#13).

Branch protection and rulesets are both unavailable on this private repo
without GitHub Pro -- measured directly, not inferred from docs:

    $ gh api repos/jonhill90/agent-supervisor/branches/main/protection
    403 "Upgrade to GitHub Pro or make this repository public..."
    $ gh api repos/jonhill90/agent-supervisor/rulesets
    403 (same)
    $ gh api repos/jonhill90/agent-supervisor/rulesets -X POST -f name=x ...
    403 (same -- creating one is refused too, not just listing)

So GitHub enforces nothing here: a red run does not block `gh pr merge`,
it only informs whoever is looking, and "CI gates this repo" is a
sentence about convention, not a mechanism. This module is the mechanism
-- a hard refusal a merge wrapper (`merge-pr.sh`) must consult, in the
actor rather than the platform. It is still weaker than real branch
protection: nothing stops a caller from running `gh pr merge` directly
and skipping this module entirely. That gap is not closeable without
Pro or a public repo, and is stated here rather than left implicit.

THE SHA THIS CHECKS, exactly -- `evaluate()` takes only `repo` and
`number`, never a caller-supplied head SHA. It fetches `headRefOid`
itself, immediately before reading checks, and reads checks for that
exact SHA via `commits/{sha}/check-runs` and `commits/{sha}/status`
(SHA-scoped endpoints, not `gh pr view`'s `statusCheckRollup`, which
carries no SHA a caller can compare). A caller that fetched a PR
snapshot earlier -- the supervisor's sensor.py does this every tick --
and passed that snapshot's SHA in would reintroduce exactly the bug this
exists to close: check green for an OLDER commit than head, because the
push that invalidated it happened after the snapshot was taken and
before the merge. Refusing the parameter closes that path structurally
instead of relying on every caller to remember to re-fetch.

Even the SHA-scoped `check-runs` response is not trusted blindly: each
run's own `head_sha` field is compared against the SHA just fetched
(`_refuse_on_sha_mismatch`). GitHub's API should never return a run for
the wrong commit given a SHA-scoped URL -- this defends against a proxy,
a cache, or a future bug doing exactly that, and is the one comparison
the "mutate and confirm it goes red" step in the brief targets.

Absent is not pending (#149's distinction, inherited from verdict.py's
GithubReviewVerdictSource / LedgerVerdictSource). A SHA with zero check
runs and zero legacy statuses is refused, not treated as "nothing to
block on" -- a push that raced past CI, or a workflow that never
triggered, must not read the same as a green run.

RE-RUNS (#269) -- `commits/{sha}/check-runs` returns every run of every
check for that SHA, not just the newest: a re-run of `test` after a flaky
failure leaves both the old failing run and the new passing one in the
list. `gh pr checks` and GitHub's PR UI evaluate only the latest run per
check name; `_latest_per_name` does the same here, keyed on `completed_at`
falling back to `started_at` for a run still in flight. This is
latest-per-name, never best-per-name -- if the newest run of a check
failed and an older run of the same check passed, that is red. Only
narrowing what counts as "the check", never widening what counts as
green.

STALE `gate` SUPERSEDED BY `ui-evidence` (#518, surfaced by #516) --
`_latest_per_name` collapses re-runs of the SAME name, but `ui-evidence.yml`'s
`gate` job (push-triggered) and the separately-named `ui-evidence`
check-run (issue_comment-triggered, satisfying a follow-up-comment #468
evidence pattern) are two DIFFERENT names for what is really the same
requirement -- `_latest_per_name` has no relationship between them to
collapse, so a `gate` job that only ever runs on push stays a permanent
FAILURE even after the PR is genuinely green.
`_gate_superseded_by_ui_evidence` is the one narrow exception to "only
narrowing what counts as green" above: a FAILURE `gate` run is excluded
from the failing list ONLY when a same-SHA `ui-evidence` run is itself
green and completed strictly later than the `gate` failure. This is named
explicitly for this one relationship, not a general "an unrelated green
check clears an unrelated red one" rule.
"""

from __future__ import annotations

import argparse
import json
import subprocess
import sys


def subprocess_runner(command):
    return subprocess.run(command, check=True, capture_output=True, text=True, timeout=30).stdout


GREEN_CONCLUSIONS = ("success", "neutral", "skipped")


def _run_sort_key(run):
    # `completed_at` is missing while a run is still queued/in_progress --
    # fall back to `started_at` so an in-progress run still sorts by when it
    # started, rather than being treated as older than every finished run
    # (which would let a stale success win the "latest" slot). Both are
    # ISO-8601 UTC timestamps ("...Z"), so lexical order matches chronological
    # order without parsing them.
    return run.get("completed_at") or run.get("started_at") or ""


# agent-supervisor#380: a merge gate that only ever prints "not green" makes
# `cancelled` (will never go green without a re-run) read identically to
# `in_progress`/`conclusion: null` (may still go green on its own). Whoever
# reads this reason -- a human or a lane deciding whether to wait or
# re-dispatch -- needs that distinction spelled out, even though the
# decision itself (refuse) is the same either way.
def _describe_failing_run(run):
    name = run.get("name", "?")
    if run.get("status") == "completed" and run.get("conclusion") == "cancelled":
        return f"{name} (cancelled -- will not go green without a re-run)"
    if run.get("status") != "completed":
        return f"{name} ({run.get('status', 'unknown')} -- may still go green)"
    return f"{name} ({run.get('conclusion', 'unknown')})"


def _latest_per_name(check_runs):
    """Re-runs of the same check (e.g. `test`) land as separate entries in
    `check_runs` -- GitHub's own PR UI and `gh pr checks` both evaluate only
    the newest one per name. Do the same here: a stale failing run must not
    keep refusing once a newer run of the same check has succeeded, and
    conversely a fresh failure must not be shadowed by an older success."""
    latest_by_name = {}
    for run in check_runs:
        name = run.get("name", "?")
        current = latest_by_name.get(name)
        if current is None or _run_sort_key(run) >= _run_sort_key(current):
            latest_by_name[name] = run
    return list(latest_by_name.values())


def _gate_superseded_by_ui_evidence(latest_runs):
    """agent-supervisor#518 / #516: `ui-evidence.yml`'s `gate` job runs only
    on `pull_request` (push) and is never re-triggered when a PR satisfies
    UI evidence via a follow-up comment -- the #468 pattern this repo's own
    docs describe as valid. That leaves a permanently stale FAILURE `gate`
    run sitting at the head SHA even after the PR is genuinely green,
    because the separately-named, `issue_comment`-triggered `ui-evidence`
    check-run (workflow "Validate") that actually reflects reality lands
    under a different name and `_latest_per_name` has no relationship
    between the two names to collapse.

    This is narrowly the one stale/superseding relationship named in
    agent-supervisor#518 -- a FAILURE `gate` run superseded ONLY by a
    same-SHA `ui-evidence` run that is itself green AND completed strictly
    later than the stale failure. It does not become a general "ignore old
    failures" rule:
      - no same-SHA `ui-evidence` run at all -> not superseded, still refuse
      - the `ui-evidence` run exists but is not itself green -> not
        superseded, still refuse (a failing superseding check clears nothing)
      - the `ui-evidence` run completed BEFORE the `gate` failure (i.e. the
        `gate` failure is a real, later failure, not a stale earlier one)
        -> not superseded, still refuse
    """
    gate = next((run for run in latest_runs if run.get("name") == "gate"), None)
    if gate is None:
        return False
    if not (gate.get("status") == "completed" and gate.get("conclusion") == "failure"):
        return False

    ui_evidence = next((run for run in latest_runs if run.get("name") == "ui-evidence"), None)
    if ui_evidence is None:
        return False
    if not (ui_evidence.get("status") == "completed" and ui_evidence.get("conclusion") in GREEN_CONCLUSIONS):
        return False

    return _run_sort_key(ui_evidence) > _run_sort_key(gate)


class CiGate:
    def __init__(self, runner=None):
        self.runner = runner or subprocess_runner

    def _head_sha(self, repo, number):
        raw = self.runner(["gh", "pr", "view", str(number), "--repo", repo, "--json", "headRefOid"])
        payload = json.loads(raw)
        sha = payload.get("headRefOid") if isinstance(payload, dict) else None
        if not isinstance(sha, str) or not sha:
            raise ValueError("gh did not return a head SHA for this PR")
        return sha

    def _check_runs(self, repo, sha):
        raw = self.runner(["gh", "api", f"repos/{repo}/commits/{sha}/check-runs", "--jq", ".check_runs"])
        runs = json.loads(raw)
        if not isinstance(runs, list):
            raise ValueError("gh returned malformed check-runs for this SHA")
        return runs

    def _combined_status(self, repo, sha):
        raw = self.runner(["gh", "api", f"repos/{repo}/commits/{sha}/status", "--jq", "{statuses: .statuses}"])
        payload = json.loads(raw)
        statuses = payload.get("statuses") if isinstance(payload, dict) else None
        if not isinstance(statuses, list):
            raise ValueError("gh returned malformed commit status for this SHA")
        return statuses

    def evaluate(self, *, repo, number):
        try:
            sha = self._head_sha(repo, number)
        except Exception as error:
            return {"decision": "refuse", "sha": None, "reason": f"could not read PR head: {error}"}

        try:
            check_runs = self._check_runs(repo, sha)
            statuses = self._combined_status(repo, sha)
        except Exception as error:
            return {"decision": "refuse", "sha": sha, "reason": f"could not read checks for {sha}: {error}"}

        for run in check_runs:
            run_sha = run.get("head_sha") if isinstance(run, dict) else None
            if run_sha != sha:
                return {
                    "decision": "refuse",
                    "sha": sha,
                    "reason": (
                        f"check run {run.get('name', '?') if isinstance(run, dict) else '?'} "
                        f"reports head_sha {run_sha!r}, not PR head {sha}"
                    ),
                }

        if not check_runs and not statuses:
            return {"decision": "refuse", "sha": sha, "reason": f"no checks reported for {sha}"}

        latest_runs = _latest_per_name(check_runs)
        gate_superseded = _gate_superseded_by_ui_evidence(latest_runs)
        failing_runs = [
            _describe_failing_run(run)
            for run in latest_runs
            if not (run.get("status") == "completed" and run.get("conclusion") in GREEN_CONCLUSIONS)
            and not (gate_superseded and run.get("name") == "gate")
        ]
        failing_statuses = [s.get("context", "?") for s in statuses if s.get("state") != "success"]
        failing = failing_runs + failing_statuses
        if failing:
            return {
                "decision": "refuse",
                "sha": sha,
                "reason": f"check(s) not green at {sha}: {', '.join(failing)}",
            }

        return {"decision": "allow", "sha": sha, "reason": f"all checks green at {sha}"}


def main(argv=None):
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("--repo", required=True, help="owner/name")
    parser.add_argument("--number", type=int, required=True)
    args = parser.parse_args(argv)

    gate = CiGate()
    result = gate.evaluate(repo=args.repo, number=args.number)
    print(json.dumps(result))
    return 0 if result["decision"] == "allow" else 1


if __name__ == "__main__":
    sys.exit(main())
